### PR TITLE
File upload

### LINKED ISSUE #2: Add middleware to upload PDF
Just navigate through the application and add the functionality to upload PDF using any 3rd party library.

### PR DESCRIPTION
**Pull request by AbhishekJadhav2002**
Completing the task no. 1 mentioned in [#2](https://github.com/surya7765/shark-tank/issues/2#issuecomment-1292350697)
Uploaded files get saved in the `backend/uploads` directory.
> But if the cloud deployment of the server restarts, then this local files will be **lost**. That's why it is more logical to save this files in the database or cloud storage, MongoDB in this case, which is next task to accomplish (see [#2](https://github.com/surya7765/shark-tank/issues/2#issuecomment-1292350697)). 

##### _fixes #2_

### CODE REVIEW

**Comment by surya7765** [comment]
> Completing the task no. 1 mentioned in [#2](https://github.com/surya7765/shark-tank/issues/2#issuecomment-1292350697) Uploaded files get saved in the `backend/uploads` directory.
> 
> > But if the cloud deployment of the server restarts, then this local files will be **lost**. That's why it is more logical to save this files in the database or cloud storage, MongoDB in this case, which is next task to accomplish (see [#2](https://github.com/surya7765/shark-tank/issues/2#issuecomment-1292350697)).
> 
> ##### _fixes #2_

It would be better if these files get stored in any cloud services. It'll be great while deployment.

**Comment by AbhishekJadhav2002** [comment]
Okay 👍🏻
What cloud storage service to use ?
([This](https://qr.ae/pvlXvZ) are some service providers)

**Comment by surya7765** [comment]
you can use any but it should be secure. And I don't think the way u have added the code for the file upload will work anyway. I have already tried so many things there. I could have used AWS not a problem. but the only thing is that it needs to be uploaded not only from postman i.e. over API but also from the front-end (in this case ReactJs).

**Comment by AbhishekJadhav2002** [comment]
> you can use any but it should be secure. And I don't think the way u have added the code for the file upload will work anyway. I have already tried so many things there. I could have used AWS not a problem. but the only thing is that it needs to be uploaded not only from postman i.e. over API but also from the front-end (in this case ReactJs).

-----
Yes I agree, I guess the workflow can be like,

**->** first user post their resume from frontend **->** which is captured on our backend API **->** then we send that file to cloud storage provider and get the link for that file from them **->** and save that link into our own MongoDB document.

Or another optioin would be like,
if we do not want third party cloud storage providers then we can store the files in our own MongoDB document. For this option we also need a extra rendering mechanism which can read the binary data and render them as PDF.

------

I will experiment with different service providers for now.

**Comment by AbhishekJadhav2002** [comment]
@surya7765 I have tried uploading resume and saved those on [Cloudinary](https://cloudinary.com/). And now the files are stored in our MongoDB collection as `"resume": "URL_to_that_file"` returned by [Cloudinary Upload API](https://cloudinary.com/documentation/node_quickstart).

For pricing part, I have gone with Free plan and it provided me this [`CLOUDINARY_URL`](https://github.com/AbhishekJadhav2002/shark-tank/blob/a65cb10a4c4f21dda272a5c9acf7ac11135631bf/backend/nodemon.json#L3).

File: backend/validators/UserValidation.js [hunk 1]
```diff
@@ -21,16 +21,16 @@ validators.validateEmail = (email) => {
 };
 
 validators.validatePassword = (pass) => {
-  if (pass.length < 8) {
-    let err = new Error("Password should be atleast 8 characters long");
+  if (pass?.length < 8) {
```
Comment (surya7765) [inline backend/validators/UserValidation.js:24]: what you did is correct but it doesn't work while working with ReactJs or Postman

File: backend/models/JobPost.js [hunk 2]
```diff
@@ -1,73 +1,78 @@
...
-    type: String,
-    required: [true, "Industry type is required"],
-  },
-  searchKeywords: {
-    type: String,
-    required: [true, "Search Keywords is required"],
-  },
-});
-
+const JobPostSchema = new mongoose.Schema(
```
Comment (surya7765) [inline backend/models/JobPost.js:3]: You changed so many things unnecessarily.

File: backend/middleware/UploadPDF.js [hunk 3]
```diff
@@ -1,12 +1,10 @@
 import path from "path";
 import multer from "multer";
 const __dirname = path.resolve();
-console.log(path.join(__dirname, "./uploads/"));
+// console.log(path.join(__dirname, "./uploads/"));
 
 var storage = multer.diskStorage({
-  destination: function (req, file, cb) {
```
Comment (surya7765) [inline backend/middleware/UploadPDF.js:7]: I don't think it's gonna work this way.

**Comment by surya7765** [comment]
Bro, do one thing once again u fork this repo and change only necessary things, and then give a PR. The code should be working. Just finish it by today or tmrw and I'll approve it for hacktoberfest.

**Comment by AbhishekJadhav2002** [comment]
Actually what has happened is that, from my [initial commit](https://github.com/surya7765/shark-tank/pull/5/commits/77f906e0b92d721f6fc77d08ecb79604aa627abe), I took `jobposts` as reference branch because it was having most of the working code and then continued working from that.
But, while creating this PR, I compared my work with `master` which is having less code than my initial point of reference that is `jobposts` branch.

**Comment by AbhishekJadhav2002** [comment]
> Bro, do one thing once again u fork this repo and change only necessary things, and then give a PR. The code should be working. Just finish it by today or tmrw and I'll approve it for hacktoberfest.

Ok then I'll do this, and this time from `master` branch 😅

**Comment by surya7765** [comment]
Do not change any code of other folders. You can change only if it's needed. As I have not committed so many things since 3 months, in my local environment there's too much code I have written since then.

**Comment by AbhishekJadhav2002** [comment]
> Do not change any code of other folders. You can change only if it's needed. As I have not committed so many things since 3 months, in my local environment there's too much code I have written since then.

Yes, I have created another PR #8, as you instructed.
I guess this PR can be closed now 😄

**Comment by surya7765** [comment]
Another PR accepted.

**Pull request closed without merging**